src/impl_owned_array.rs

```diff
@@ -61,6 +61,55 @@ impl<A, D> Array<A, D>
 where
     D: Dimension,
 {
+    /// Returns the offset (in units of `A`) from the start of the raw data to
+    /// the first element, or `None` if the array is empty.
+    ///
+    /// In other words, after converting the array to `Vec<A>` with
+    /// [`.into_raw_vec()`](Self::into_raw_vec), this is the offset from the
+    /// start of the `Vec` to the first element of the array.
+    ///
+    /// ```
+    /// use ndarray::{Array2, Axis};
+    ///
+    /// let mut arr: Array2<f64> = Array2::zeros([3, 4]);
+    /// arr.slice_axis_inplace(Axis(0), (1..).into());
+    /// arr[[0, 0]] = 5.;
+    ///
+    /// let offset = arr.offset_to_first_elem().unwrap();
+    /// assert_eq!(offset, 4);
+    ///
+    /// let v = arr.into_raw_vec();
+    /// assert_eq!(v[offset], 5.);
+    /// ```
+    ///
+    /// In the case of zero-sized elements, the offset is somewhat meaningless.
+    /// For convenience, an offset will be returned such that all indices
+    /// computed using the offset, shape, and strides will be in-bounds for the
+    /// `Vec<A>` returned by [`.into_raw_vec()`](Self::into_raw_vec). Note that
+    /// this offset won't necessarily be the same as the offset for an array of
+    /// nonzero-sized elements sliced in the same way.
+    ///
+    /// ```
+    /// use ndarray::{Array2, Axis};
+    ///
+    /// let mut arr: Array2<()> = Array2::from_elem([3, 4], ());
+    /// arr.slice_axis_inplace(Axis(0), (1..).into());
+    /// let offset = arr.offset_to_first_elem().unwrap();
+    /// assert_eq!(offset, 0);
+    /// ```
+    pub fn offset_to_first_elem(&self) -> Option<usize> {
+        if self.is_empty() {
+            return None;
+        }
+        let offset = if std::mem::size_of::<A>() == 0 {
+            -dimension::offset_from_ptr_to_memory(&self.dim, &self.strides)
```

Comment:
can't we use this dimension function in both cases?

Reply:
No, because the shape and strides are not sufficient information. Consider, for example:

```rust
let a = Array2::<f64>::zeros([3, 4]).slice_move(s![1.., ..]);
let b = Array2::<f64>::zeros([2, 4]);
assert_eq!(a.shape(), b.shape());
assert_eq!(a.strides(), b.strides());
assert_ne!(
    unsafe { a.as_ptr().offset_from(a.into_raw_vec().as_ptr()) },
    unsafe { b.as_ptr().offset_from(b.into_raw_vec().as_ptr()) },
);
```

`a` and `b` have the same shape and strides but different offsets.

The only reason for calling `dimension::offset_from_ptr_to_memory` here is to get an offset such that indices computed with `offset + index[0]*strides[0] + index[1]*strides[1] + ...` will be in-bounds for the result of `.into_raw_vec()`. We could return value return `None` or `Some(0)` for the zero-sized element case, but then it's more likely that the caller will have to explicitly deal with the zero-sized element case too.

Reply:
Oh right. Thanks! That's the three "starts" of the array in memory again...

1. start of memory allocation
2. lowest address of an element in the array
3. address of logically first element in the array..

I guess if we try and find good names for these (again). I wouldn't mind allocation head, memory head, logical head or something along those lines. Maybe "memory" is too vague to be useful(?)

Reply:
I like "allocation ptr" and "logical ptr/offset". "Memory head" to refer to the lowest address of any element is a bit confusing, IMO. I'd rather say "low addr ptr/offset". Fortunately, there aren't many times when that pointer/offset is important, anyway, so a concise name for it probably isn't a big concern. Pretty much the only things I could see the "low addr head" being useful for are the `drop_unreachable_elements` method or a method to "compact" an array (remove unreachable elements).

Reply:
> I mixed up (1) and (2)...

Fwiw, this isn't quite true. The mixup was assuming that (1) and (3) were the same. The `dimension::offset_from_ptr_to_memory` function actually returns the offset from the "logical pointer" (3) to the "lowest reachable address pointer" (2). This makes me realize that we should really rename `dimension::offset_from_ptr_to_memory` to `dimension::offset_from_logical_ptr_to_low_addr_ptr`. (Actually, I'd prefer to have a `dimension::offset_from_low_addr_ptr_to_logical_ptr` function instead so that the return value is always nonnegative.)

Reply:
I don't think "logical ptr" is very descriptive either. Offset >= 0 sounds good.

Reply:
I don't mind "logical ptr", but we can think of another name. Maybe "zero idx ptr" / "zero index ptr"?